examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+                             );
+
+    void vmult (BlockVector<double>       &dst,
+                const BlockVector<double> &src) const;
+
+  private:
+    const double gamma;
+    const double viscosity;
+    const BlockSparseMatrix<double> &stokes_matrix;
+    const SparseMatrix<double> &pressure_mass_matrix;
+    const PreconditionerMp &mp_preconditioner;
+    SparseDirectUMFPACK A_direct;
+  };
+
+  template <class PreconditionerMp>
+  BlockSchurPreconditioner<PreconditionerMp>::
+  BlockSchurPreconditioner (double                           gamma,
+                            double                           viscosity,
+                            const BlockSparseMatrix<double>  &S,
+                            const SparseMatrix<double>       &P,
+                            const PreconditionerMp           &Mppreconditioner)
+    :
+    gamma                (gamma),
+    viscosity            (viscosity),
+    stokes_matrix        (S),
+    pressure_mass_matrix (P),
+    mp_preconditioner    (Mppreconditioner)
+  {
+    A_direct.initialize(stokes_matrix.block(0,0));
+  }
+
+  template <class PreconditionerMp>
+  void
+  BlockSchurPreconditioner<PreconditionerMp>::
+  vmult (BlockVector<double>       &dst,
+         const BlockVector<double> &src) const
+  {
+    Vector<double> utmp(src.block(0));
+
+    {
+      SolverControl solver_control(1000, 1e-6 * src.block(1).l2_norm());
+      SolverCG<>    cg (solver_control);
+
+      dst.block(1) = 0.0;
+      cg.solve(pressure_mass_matrix,
+               dst.block(1), src.block(1),
+               mp_preconditioner);
+      dst.block(1) *= -(viscosity+gamma);
+    }
+
+    {
+      stokes_matrix.block(0,1).vmult(utmp, dst.block(1));
+      utmp*=-1.0;
+      utmp+=src.block(0);
+    }
+
+    A_direct.vmult (dst.block(0), utmp);
+  }
+
+  // @sect3{StationaryNavierStokes class implementation}
+  // @sect4{StationaryNavierStokes::StationaryNavierStokes}
+  // The constructor of this class looks very similar to the one in step-22. The only
+  // difference is the viscosity and the AL coefficient gamma.
+  //
+
+  template <int dim>
+  StationaryNavierStokes<dim>::StationaryNavierStokes(const unsigned int degree)
+    :
+    viscosity(1.0/7500.0),
+    gamma(1.0),
+    degree(degree),
+    triangulation(Triangulation<dim>::maximum_smoothing),
+    fe(FE_Q<dim>(degree+1), dim,
+       FE_Q<dim>(degree),   1),
+    dof_handler(triangulation)
+  {}
+
+  // @sect4{StationaryNavierStokes::setup_dof}
+  // This function provide the information of DoF and initialize the constraints
+  // on current mesh.
+  template <int dim>
+  void StationaryNavierStokes<dim>::setup_dof()
+  {
+    system_matrix.clear();
+    pressure_mass_matrix.clear();
+
+    // The first step is to associate DoFs with a given mesh.
+    dof_handler.distribute_dofs (fe);
+
+    // We renumber the components to have all velocity DoFs come before the pressure
+    // DoFs to be able to use the block preconditioner.
+    //
```

Comment:
How about: "We renumber the components to have all velocity DoFs come before the pressure DoFs to be able to split the solution vector in two blocks which are separately accessed in the block preconditioner."